pkg/envoy/xds/ack.go

```diff
@@ -202,7 +246,55 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
 
 		// Completion didn't match or is still waiting for some ACKs. Keep it
 		// in the pending list.
-		remainingCompletions = append(remainingCompletions, comp)
+		remainingCompletions[comp] = pending
+	}
+
+	m.pendingCompletions = remainingCompletions
+}
+
+// HandleResourceVersionNack handles NACKs by calling callbacks of pending completions, if set.
+func (m *AckingResourceMutatorWrapper) HandleResourceVersionNack(version uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string, detail string) {
+	ackLog := log.WithFields(logrus.Fields{
+		logfields.XDSVersionInfo: version,
+		logfields.XDSClientNode:  node,
+		logfields.XDSTypeURL:     typeURL,
+	})
+
+	nodeID, err := m.nodeToID(node)
+	if err != nil {
+		// Ignore NACKs from unknown or misconfigured nodes which have invalid
+		// node identifiers.
+		ackLog.Warning("Invalid ID in Node identifier; ignoring NACK")
+		return
+	}
+
+	m.locker.Lock()
+	defer m.locker.Unlock()
+
+	remainingCompletions := make(map[*completion.Completion]*pendingCompletion, len(m.pendingCompletions))
+
+	for comp, pending := range m.pendingCompletions {
+		if comp.Context().Err() != nil {
+			// Completion was canceled or timed out.
+			// Remove from pending list.
+			ackLog.Debugf("completion context was canceled: %v", pending)
+			comp.Complete(context.Canceled)
```

Comment:
To be consistent with what you do in the wait group:
```
comp.Complete(comp.Context().Err())
```

Reply:
With the changes due to this review the Complete() call is redundant and hence removed.